nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/ConsumePulsar.java

```diff
@@ -0,0 +1,141 @@
+            }
+
+            if (context.getProperty(ASYNC_ENABLED).asBoolean()) {
+                consumeAsync(consumer, context, session);
+                handleAsync(consumer, context, session);
+            } else {
+                consume(consumer, context, session);
+            }
+        } catch (PulsarClientException e) {
+            getLogger().error("Unable to consume from Pulsar Topic ", e);
+            context.yield();
+            throw new ProcessException(e);
+        }
+    }
+
+    private void handleAsync(final Consumer<byte[]> consumer, ProcessContext context, ProcessSession session) {
+        try {
+            Future<Message<byte[]>> done = getConsumerService().poll(50, TimeUnit.MILLISECONDS);
+
+            if (done != null) {
+               Message<byte[]> msg = done.get();
+
+               if (msg != null) {
+                  FlowFile flowFile = null;
+                  final byte[] value = msg.getData();
+                  if (value != null && value.length > 0) {
+                      flowFile = session.create();
+                      flowFile = session.write(flowFile, out -> {
+                          out.write(value);
+                      });
+
+                     session.getProvenanceReporter().receive(flowFile, getPulsarClientService().getPulsarBrokerRootURL() + "/" + consumer.getTopic());
+                     session.transfer(flowFile, REL_SUCCESS);
+                     session.commit();
+                  }
+                  // Acknowledge consuming the message
+                  getAckService().submit(new Callable<Object>() {
+                      @Override
+                      public Object call() throws Exception {
+                         return consumer.acknowledgeAsync(msg).get();
+                      }
+                   });
+              }
+            } else {
+               // Check if the processor is stopped and if so gracefully stop processing.
+            }
+
+        } catch (InterruptedException | ExecutionException e) {
+            getLogger().error("Trouble consuming messages ", e);
+        }
+    }
+
+    private void consume(Consumer<byte[]> consumer, ProcessContext context, ProcessSession session) throws PulsarClientException {
+        try {
+            /* If we do not provide a timeout value, this call will block indefinitely, which will
+             * cause the processor to hang when we try to stop it consuming from an empty topic.
+             * Therefore, we chose a 2 second timeout to handle this situation gracefully.
+             */
+            final Message<byte[]> msg = consumer.receive(2, TimeUnit.SECONDS);
+
+            if (msg == null) {
+               return;
+            }
+
+            final byte[] value = msg.getData();
+
+            if (value != null && value.length > 0) {
+                FlowFile flowFile = session.create();
+                flowFile = session.write(flowFile, out -> {
+                    out.write(value);
```

Comment:
performance of a single flowfile per poll cycle will be poor relative to doing batched reads.  I am guessing pulsar pulls more than a single message at a time too.  We have the demarcator approach for the kafka processors and performance there is extremely high because then we're aligning nifi and the messaging systems strengths.  That said, the record processors are an even better path for this so maybe dont worry about that for now.  not sure.